ethexe/consensus/src/connect/mod.rs

```diff
@@ -0,0 +1,129 @@
+// This file is part of Gear.
+//
+// Copyright (C) 2025 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! # "Connect-Node" Consensus Service
+//!
+//! Simple "connect-node" consensus service implementation.
+
+use crate::{
+    utils::{BatchCommitmentValidationReply, BatchCommitmentValidationRequest},
+    ConsensusEvent, ConsensusService,
+};
+use anyhow::Result;
+use ethexe_common::{ProducerBlock, SimpleBlockData};
+use ethexe_observer::BlockSyncedData;
+use ethexe_signer::SignedData;
+use futures::{stream::FusedStream, Stream};
+use gprimitives::H256;
+use std::{
+    collections::VecDeque,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+/// Consensus service which tracks the on-chain and ethexe events
+/// in order to keep the program states in local database actual.
+#[derive(Debug, Default)]
+pub struct SimpleConnectService {
+    chain_head: Option<SimpleBlockData>,
+    output: VecDeque<ConsensusEvent>,
+}
+
+impl SimpleConnectService {
+    /// Creates a new instance of `SimpleConnectService`.
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+
+impl ConsensusService for SimpleConnectService {
+    fn role(&self) -> String {
+        "Connect".to_string()
+    }
+
+    fn receive_new_chain_head(&mut self, block: SimpleBlockData) -> Result<()> {
+        self.chain_head = Some(block);
+
+        Ok(())
+    }
+
+    fn receive_synced_block(&mut self, data: BlockSyncedData) -> Result<()> {
+        let Some(block) = self.chain_head.as_ref() else {
+            self.output.push_back(ConsensusEvent::Warning(format!(
+                "Received synced block {}, but no chain-head was received yet",
+                data.block_hash
+            )));
+
+            return Ok(());
+        };
+
+        if block.hash != data.block_hash {
+            self.output.push_back(ConsensusEvent::Warning(format!(
+                "Received synced block {} is different from the expected block hash {}",
+                data.block_hash, block.hash
+            )));
+
+            return Ok(());
+        }
```

Comment:
These branches are impossible because `Observer` produce its events sequentially and in exact order

Reply:
1) service logic should not rely on other services implementations
2) Yes this service does not expect this behaviour in normal situation (but supports it) and so generate warnings
3) Second branch could happen when we receive in a short time two or more blocks - in this case first block would be synced after the second block chain head is received by consensus service. So, second branch could happen even for current observer and main service implementations.